File: shared/redis_client.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any, Iterable

import redis.asyncio as redis

from shared.protocol import AgentMessage, Task, WorkerRequest


@dataclass(frozen=True)
class RedisKeys:
    jobs_stream: str = "stream:jobs"
    agents_active: str = "agents:active"
    worker_request_queue: str = "queue:worker_requests"

    def job_hash(self, job_id: str) -> str:
        return f"job:{job_id}"

    def job_tasks_zset(self, job_id: str) -> str:
        return f"job:{job_id}:tasks"

    def job_next_task_id(self, job_id: str) -> str:
        return f"job:{job_id}:next_task_id"

    def task_hash(self, job_id: str, task_id: str) -> str:
        return f"task:{job_id}:{task_id}"

    def task_blocks(self, job_id: str, task_id: str) -> str:
        return f"task:{job_id}:{task_id}:blocks"

    def task_blocked_by(self, job_id: str, task_id: str) -> str:
        return f"task:{job_id}:{task_id}:blockedBy"

    def task_lock(self, job_id: str, task_id: str) -> str:
        return f"lock:task:{job_id}:{task_id}"

    def inbox(self, job_id: str, agent_name: str) -> str:
        return f"inbox:{job_id}:{agent_name}"

    def agent_hash(self, agent_id: str) -> str:
        return f"agent:{agent_id}"

    def job_events_channel(self, job_id: str) -> str:
        return f"channel:job:{job_id}:events"


class RedisCoordinator:
    """Single class handling all Redis coordination patterns."""

    def __init__(self, redis_url: str):
        self.keys = RedisKeys()
        self.pool = redis.ConnectionPool.from_url(redis_url, decode_responses=False)
        self.redis = redis.Redis(connection_pool=self.pool)

    async def close(self) -> None:
        await self.redis.aclose()
        await self.pool.disconnect(inuse_connections=True)

    # --- Job metadata (Hash) ---
    async def upsert_job(self, job_id: str, data: dict[str, Any]) -> None:
        mapping = {
            k: json.dumps(v).encode()
            if isinstance(v, (dict, list))
            else str(v).encode()
            for k, v in data.items()
        }
        await self.redis.hset(self.keys.job_hash(job_id), mapping=mapping)

    async def get_job(self, job_id: str) -> dict[str, Any] | None:
        raw = await self.redis.hgetall(self.keys.job_hash(job_id))
        if not raw:
            return None
        decoded: dict[str, Any] = {}
        for k, v in raw.items():
            key = k.decode()
            val = v.decode()
            if key in {"config", "directives"}:
                try:
                    decoded[key] = json.loads(val)
                    continue
                except Exception:
                    pass
            decoded[key] = val
        return decoded

    async def set_job_status(self, job_id: str, status: str) -> None:
        await self.redis.hset(
            self.keys.job_hash(job_id), mapping={"status": status.encode()}
        )
        await self.publish_job_event(job_id, {"type": "job_status", "status": status})

    # --- Jobs (Stream) ---
    async def submit_job(self, job: dict[str, Any]) -> str:
        job_id = str(job.get("job_id") or job.get("id") or "")
        payload = {"data": json.dumps(job)}
        await self.redis.xadd(self.keys.jobs_stream, payload)
        return job_id

    async def ensure_consumer_group(self, group: str) -> None:
        try:
            await self.redis.xgroup_create(
                self.keys.jobs_stream, group, id="0", mkstream=True
            )
        except redis.ResponseError as e:
            if "BUSYGROUP" not in str(e):
                raise

    async def read_jobs(
        self,
        group: str,
        consumer: str,
        block_ms: int = 5000,
        count: int = 1,
    ):
        return await self.redis.xreadgroup(
            groupname=group,
            consumername=consumer,
            streams={self.keys.jobs_stream: ">"},
            count=count,
            block=block_ms,
        )

    async def ack_job(self, group: str, entry_id: str) -> None:
        await self.redis.xack(self.keys.jobs_stream, group, entry_id)

    # --- Inboxes (List) ---
    async def send_message(
        self, job_id: str, target_agent: str, message: AgentMessage
    ) -> None:
        await self.redis.lpush(
            self.keys.inbox(job_id, target_agent), message.model_dump_json().encode()
        )

    async def wait_for_message(
        self, job_id: str, agent_name: str, timeout: int = 0
    ) -> AgentMessage | None:
        result = await self.redis.brpop(
            self.keys.inbox(job_id, agent_name), timeout=timeout
        )
        if not result:
            return None
        _, raw = result
        if isinstance(raw, bytes):
            raw = raw.decode("utf-8")
        return AgentMessage.model_validate_json(raw)

    async def broadcast(
        self, job_id: str, message: AgentMessage, agent_names: Iterable[str]
    ) -> None:
        pipe = self.redis.pipeline()
        payload = message.model_dump_json().encode()
        for name in agent_names:
            pipe.lpush(self.keys.inbox(job_id, name), payload)
        await pipe.execute()

    # --- Tasks (Hashes/Sets/ZSet + Locks) ---
    async def create_task(self, job_id: str, task: Task) -> str:
        task_id = str(await self.redis.incr(self.keys.job_next_task_id(job_id)))
        task.id = task_id
        mapping: dict[bytes, bytes] = {
            b"id": task_id.encode(),
            b"subject": task.subject.encode(),
            b"description": task.description.encode(),
            b"activeForm": task.activeForm.encode(),
            b"owner": task.owner.encode(),
            b"status": task.status.encode(),
        }
        await self.redis.hset(self.keys.task_hash(job_id, task_id), mapping=mapping)
        await self.redis.zadd(
            self.keys.job_tasks_zset(job_id), {task_id.encode(): time.time()}
        )

        for downstream in task.blocks:
            await self.redis.sadd(
                self.keys.task_blocks(job_id, task_id), downstream.encode()
            )
            await self.redis.sadd(
                self.keys.task_blocked_by(job_id, downstream), task_id.encode()
            )

        for upstream in task.blockedBy:
            await self.redis.sadd(
                self.keys.task_blocked_by(job_id, task_id), upstream.encode()
            )
            await self.redis.sadd(
                self.keys.task_blocks(job_id, upstream), task_id.encode()
            )

        await self.publish_job_event(
            job_id, {"type": "task_created", "task_id": task_id}
        )
        return task_id

    async def get_task(self, job_id: str, task_id: str) -> Task | None:
        data = await self.redis.hgetall(self.keys.task_hash(job_id, task_id))
        if not data:
            return None
        decoded = {k.decode(): v.decode() for k, v in data.items()}
        blocked_by = await self.redis.smembers(
            self.keys.task_blocked_by(job_id, task_id)
        )
        blocks = await self.redis.smembers(self.keys.task_blocks(job_id, task_id))
        decoded["blockedBy"] = sorted([b.decode() for b in blocked_by])
        decoded["blocks"] = sorted([b.decode() for b in blocks])
        return Task.model_validate(decoded)

    async def list_tasks(self, job_id: str) -> list[Task]:
        task_ids = await self.redis.zrange(self.keys.job_tasks_zset(job_id), 0, -1)
        tasks: list[Task] = []
        for raw in task_ids:
            tid = raw.decode() if isinstance(raw, bytes) else str(raw)
            t = await self.get_task(job_id, tid)
            if t:
                tasks.append(t)
        return tasks

    async def claim_task(
        self, job_id: str, task_id: str, agent_id: str, ttl_seconds: int = 300
    ) -> bool:
        claimed = await self.redis.set(
            self.keys.task_lock(job_id, task_id),
            agent_id.encode(),
            nx=True,
            ex=ttl_seconds,
        )
        if claimed:
            await self.redis.hset(
                self.keys.task_hash(job_id, task_id),
                mapping={"status": b"in_progress", "owner": agent_id.encode()},
            )
            await self.publish_job_event(
                job_id,
                {"type": "task_claimed", "task_id": task_id, "agent_id": agent_id},
            )
        return bool(claimed)

    async def _release_task_lock(
        self, job_id: str, task_id: str, agent_id: str
    ) -> None:
        # Atomic check-and-delete.
        script = (
            "if redis.call('get',KEYS[1])==ARGV[1] then "
            "return redis.call('del',KEYS[1]) else return 0 end"
        )
        await self.redis.eval(script, 1, self.keys.task_lock(job_id, task_id), agent_id)

    async def complete_task(
        self, job_id: str, task_id: str, agent_id: str, result: str = ""
    ) -> None:
        await self.redis.hset(
            self.keys.task_hash(job_id, task_id),
            mapping={
                "status": b"completed",
                "result": result.encode(),
                "owner": agent_id.encode(),
            },
        )
        await self._release_task_lock(job_id, task_id, agent_id)

        # Auto-unblock downstream tasks.
        downstream = await self.redis.smembers(self.keys.task_blocks(job_id, task_id))
        for raw in downstream:
            tid = raw.decode() if isinstance(raw, bytes) else str(raw)
            await self.redis.srem(
                self.keys.task_blocked_by(job_id, tid), task_id.encode()
            )

        await self.redis.publish(
            "channel:tasks:completed",
            json.dumps(
                {
                    "job_id": job_id,
                    "task_id": task_id,
                    "agent_id": agent_id,
                    "status": "completed",
                    "result": result,
                }
            ).encode(),
        )
        await self.publish_job_event(
            job_id, {"type": "task_completed", "task_id": task_id, "agent_id": agent_id}
        )

    async def fail_task(
        self, job_id: str, task_id: str, agent_id: str, error: str
    ) -> None:
        await self.redis.hset(
            self.keys.task_hash(job_id, task_id),
            mapping={
                "status": b"failed",
                "result": error.encode(),
                "owner": agent_id.encode(),
            },
        )
        await self._release_task_lock(job_id, task_id, agent_id)
        await self.redis.publish(
            "channel:tasks:completed",
            json.dumps(
                {
                    "job_id": job_id,
                    "task_id": task_id,
                    "agent_id": agent_id,
                    "status": "failed",
                    "result": error,
                }
            ).encode(),
        )
        await self.publish_job_event(
            job_id,
            {
                "type": "task_failed",
                "task_id": task_id,
                "agent_id": agent_id,
                "error": error,
            },
        )

    async def get_unblocked_tasks(self, job_id: str) -> list[Task]:
        task_ids = await self.redis.zrange(self.keys.job_tasks_zset(job_id), 0, -1)
        unblocked: list[Task] = []
        for raw in task_ids:
            tid = raw.decode() if isinstance(raw, bytes) else str(raw)
            data = await self.redis.hgetall(self.keys.task_hash(job_id, tid))
            if not data:
                continue
            status = (data.get(b"status") or b"").decode()
            owner = (data.get(b"owner") or b"").decode()
            if status != "pending" or owner:
                continue
            blockers = await self.redis.smembers(self.keys.task_blocked_by(job_id, tid))
            if blockers:
                continue
            task = await self.get_task(job_id, tid)
            if task:
                unblocked.append(task)
        return unblocked

    # --- Agent registry (Hash + ZSet) ---
    async def register_agent(self, agent_id: str, agent_data: dict[str, Any]) -> None:
        mapping = {k: str(v).encode() for k, v in agent_data.items()}
        await self.redis.hset(self.keys.agent_hash(agent_id), mapping=mapping)
        await self.redis.zadd(self.keys.agents_active, {agent_id.encode(): time.time()})

    async def heartbeat(
        self, agent_id: str, status: str, current_task: str = ""
    ) -> None:
        ts = str(time.time())
        await self.redis.zadd(self.keys.agents_active, {agent_id.encode(): time.time()})
        await self.redis.hset(
            self.keys.agent_hash(agent_id),
            mapping={
                "status": status.encode(),
                "current_task": current_task.encode(),
                "heartbeat_ts": ts.encode(),
            },
        )

    async def get_idle_workers(
        self, *, stale_after_seconds: int = 30, job_id: str | None = None
    ) -> list[str]:
        cutoff = time.time() - stale_after_seconds
        agent_ids = await self.redis.zrangebyscore(
            self.keys.agents_active, cutoff, "+inf"
        )
        idle: list[str] = []
        for raw in agent_ids:
            aid = raw.decode() if isinstance(raw, bytes) else str(raw)
            data = await self.redis.hgetall(self.keys.agent_hash(aid))
            role = (data.get(b"role") or b"").decode()
            status = (data.get(b"status") or b"").decode()
            agent_job_id = (data.get(b"job_id") or b"").decode()
            if role == "worker" and status == "idle":
                if job_id and agent_job_id != job_id:
                    continue
                idle.append(aid)
        return idle

    async def get_stale_agents(self, threshold_seconds: int = 30) -> list[str]:
        cutoff = time.time() - threshold_seconds
        raw = await self.redis.zrangebyscore(self.keys.agents_active, "-inf", cutoff)
        return [r.decode() if isinstance(r, bytes) else str(r) for r in raw]

    async def list_agents(self) -> list[dict[str, str]]:
        agent_ids = await self.redis.zrange(self.keys.agents_active, 0, -1)
        agents: list[dict[str, str]] = []
        for raw in agent_ids:
            aid = raw.decode() if isinstance(raw, bytes) else str(raw)
            data = await self.redis.hgetall(self.keys.agent_hash(aid))
            agents.append({k.decode(): v.decode() for k, v in data.items()})
        return agents

    async def list_job_agents(
        self, job_id: str, stale_after_seconds: int = 30
    ) -> list[str]:
        cutoff = time.time() - stale_after_seconds
        agent_ids = await self.redis.zrangebyscore(
            self.keys.agents_active, cutoff, "+inf"
        )
        result: list[str] = []
        for raw in agent_ids:
            aid = raw.decode() if isinstance(raw, bytes) else str(raw)
            data = await self.redis.hgetall(self.keys.agent_hash(aid))
            role = (data.get(b"role") or b"").decode()
            ajob = (data.get(b"job_id") or b"").decode()
            if ajob != job_id:
                continue
            if role not in {"worker", "leader"}:
                continue
            result.append(aid)
        return result

    # --- Worker requests (List) ---
    async def submit_worker_request(self, request: WorkerRequest) -> None:
        await self.redis.lpush(
            self.keys.worker_request_queue, request.model_dump_json().encode()
        )

    async def wait_for_worker_request(self, timeout: int = 0) -> WorkerRequest | None:
        result = await self.redis.brpop(
            self.keys.worker_request_queue, timeout=timeout
        )
        if not result:
            return None
        _, raw = result
        if isinstance(raw, bytes):
            raw = raw.decode("utf-8")
        return WorkerRequest.model_validate_json(raw)

    # --- Events ---
    async def publish_job_event(self, job_id: str, event: dict[str, Any]) -> None:
        try:
            await self.redis.publish(
                self.keys.job_events_channel(job_id), json.dumps(event).encode()
            )
        except Exception:
            # Pub/Sub is best-effort.
            return

File: packages/shared/src/redis-client.ts
import { Redis } from "ioredis";
import {
  AgentMessage,
  Task,
  WorkerRequest,
  type Task as TaskType,
  type AgentMessage as AgentMessageType,
  type WorkerRequest as WorkerRequestType,
} from "./protocol.js";

// ---------------------------------------------------------------------------
// Key namespace
// ---------------------------------------------------------------------------

export class RedisKeys {
  readonly jobsStream = "stream:jobs";
  readonly agentsActive = "agents:active";
  readonly workerRequestQueue = "queue:worker_requests";

  jobHash(jobId: string) {
    return `job:${jobId}`;
  }
  jobTasksZset(jobId: string) {
    return `job:${jobId}:tasks`;
  }
  jobNextTaskId(jobId: string) {
    return `job:${jobId}:next_task_id`;
  }
  taskHash(jobId: string, taskId: string) {
    return `task:${jobId}:${taskId}`;
  }
  taskBlocks(jobId: string, taskId: string) {
    return `task:${jobId}:${taskId}:blocks`;
  }
  taskBlockedBy(jobId: string, taskId: string) {
    return `task:${jobId}:${taskId}:blockedBy`;
  }
  taskLock(jobId: string, taskId: string) {
    return `lock:task:${jobId}:${taskId}`;
  }
  inbox(jobId: string, agentName: string) {
    return `inbox:${jobId}:${agentName}`;
  }
  agentHash(agentId: string) {
    return `agent:${agentId}`;
  }
  jobEventsChannel(jobId: string) {
    return `channel:job:${jobId}:events`;
  }
}

// ---------------------------------------------------------------------------
// Coordinator
// ---------------------------------------------------------------------------

export class RedisCoordinator {
  readonly keys = new RedisKeys();
  readonly redis: Redis;

  constructor(redisUrl: string) {
    this.redis = new Redis(redisUrl);
  }

  async close(): Promise<void> {
    await this.redis.quit();
  }

  // --- Job metadata (Hash) -------------------------------------------------

  async upsertJob(
    jobId: string,
    data: Record<string, unknown>,
  ): Promise<void> {
    const flat: string[] = [];
    for (const [k, v] of Object.entries(data)) {
      flat.push(
        k,
        typeof v === "object" && v !== null ? JSON.stringify(v) : String(v),
      );
    }
    if (flat.length > 0) {
      await this.redis.hset(this.keys.jobHash(jobId), ...flat);
    }
  }

  async getJob(jobId: string): Promise<Record<string, unknown> | null> {
    const raw = await this.redis.hgetall(this.keys.jobHash(jobId));
    if (!raw || Object.keys(raw).length === 0) return null;
    const decoded: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(raw)) {
      if (k === "config" || k === "directives") {
        try {
          decoded[k] = JSON.parse(v);
          continue;
        } catch {
          // fall through
        }
      }
      decoded[k] = v;
    }
    return decoded;
  }

  async setJobStatus(jobId: string, status: string): Promise<void> {
    await this.redis.hset(this.keys.jobHash(jobId), "status", status);
    await this.publishJobEvent(jobId, { type: "job_status", status });
  }

  // --- Jobs (Stream) -------------------------------------------------------

  async submitJob(job: Record<string, unknown>): Promise<string> {
    const jobId = String(job.job_id ?? job.id ?? "");
    await this.redis.xadd(
      this.keys.jobsStream,
      "*",
      "data",
      JSON.stringify(job),
    );
    return jobId;
  }

  async ensureConsumerGroup(group: string): Promise<void> {
    try {
      await this.redis.xgroup(
        "CREATE",
        this.keys.jobsStream,
        group,
        "0",
        "MKSTREAM",
      );
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      if (!message.includes("BUSYGROUP")) throw err;
    }
  }

  async readJobs(
    group: string,
    consumer: string,
    blockMs = 5000,
    count = 1,
  ): Promise<[string, [string, string[]][]][] | null> {
    const result = await this.redis.xreadgroup(
      "GROUP",
      group,
      consumer,
      "COUNT",
      count,
      "BLOCK",
      blockMs,
      "STREAMS",
      this.keys.jobsStream,
      ">",
    );
    return result as [string, [string, string[]][]][] | null;
  }

  async ackJob(group: string, entryId: string): Promise<void> {
    await this.redis.xack(this.keys.jobsStream, group, entryId);
  }

  // --- Inboxes (List) ------------------------------------------------------

  async sendMessage(
    jobId: string,
    targetAgent: string,
    message: AgentMessageType,
  ): Promise<void> {
    await this.redis.lpush(
      this.keys.inbox(jobId, targetAgent),
      JSON.stringify(message),
    );
  }

  async waitForMessage(
    jobId: string,
    agentName: string,
    timeout = 0,
  ): Promise<AgentMessageType | null> {
    const result = await this.redis.brpop(
      this.keys.inbox(jobId, agentName),
      timeout,
    );
    if (!result) return null;
    const [, raw] = result;
    return AgentMessage.parse(JSON.parse(raw));
  }

  async broadcast(
    jobId: string,
    message: AgentMessageType,
    agentNames: string[],
  ): Promise<void> {
    const pipeline = this.redis.pipeline();
    const payload = JSON.stringify(message);
    for (const name of agentNames) {
      pipeline.lpush(this.keys.inbox(jobId, name), payload);
    }
    await pipeline.exec();
  }

  // --- Tasks (Hashes / Sets / ZSet + Locks) --------------------------------

  async createTask(jobId: string, task: TaskType): Promise<string> {
    const taskId = String(
      await this.redis.incr(this.keys.jobNextTaskId(jobId)),
    );
    task.id = taskId;

    await this.redis.hset(
      this.keys.taskHash(jobId, taskId),
      "id",
      taskId,
      "subject",
      task.subject,
      "description",
      task.description,
      "activeForm",
      task.activeForm,
      "owner",
      task.owner,
      "status",
      task.status,
    );
    await this.redis.zadd(
      this.keys.jobTasksZset(jobId),
      Date.now(),
      taskId,
    );

    for (const downstream of task.blocks) {
      await this.redis.sadd(
        this.keys.taskBlocks(jobId, taskId),
        downstream,
      );
      await this.redis.sadd(
        this.keys.taskBlockedBy(jobId, downstream),
        taskId,
      );
    }
    for (const upstream of task.blockedBy) {
      await this.redis.sadd(
        this.keys.taskBlockedBy(jobId, taskId),
        upstream,
      );
      await this.redis.sadd(
        this.keys.taskBlocks(jobId, upstream),
        taskId,
      );
    }

    await this.publishJobEvent(jobId, {
      type: "task_created",
      task_id: taskId,
    });
    return taskId;
  }

  async getTask(jobId: string, taskId: string): Promise<TaskType | null> {
    const data = await this.redis.hgetall(
      this.keys.taskHash(jobId, taskId),
    );
    if (!data || Object.keys(data).length === 0) return null;
    const blockedBy = await this.redis.smembers(
      this.keys.taskBlockedBy(jobId, taskId),
    );
    const blocks = await this.redis.smembers(
      this.keys.taskBlocks(jobId, taskId),
    );
    return Task.parse({
      ...data,
      blockedBy: blockedBy.sort(),
      blocks: blocks.sort(),
    });
  }

  async listTasks(jobId: string): Promise<TaskType[]> {
    const taskIds = await this.redis.zrange(
      this.keys.jobTasksZset(jobId),
      0,
      -1,
    );
    const tasks: TaskType[] = [];
    for (const tid of taskIds) {
      const t = await this.getTask(jobId, tid);
      if (t) tasks.push(t);
    }
    return tasks;
  }

  async claimTask(
    jobId: string,
    taskId: string,
    agentId: string,
    ttlSeconds = 300,
  ): Promise<boolean> {
    const claimed = await this.redis.set(
      this.keys.taskLock(jobId, taskId),
      agentId,
      "EX",
      ttlSeconds,
      "NX",
    );
    if (claimed) {
      await this.redis.hset(
        this.keys.taskHash(jobId, taskId),
        "status",
        "in_progress",
        "owner",
        agentId,
      );
      await this.publishJobEvent(jobId, {
        type: "task_claimed",
        task_id: taskId,
        agent_id: agentId,
      });
    }
    return claimed !== null;
  }

  private async releaseTaskLock(
    jobId: string,
    taskId: string,
    agentId: string,
  ): Promise<void> {
    const script =
      "if redis.call('get',KEYS[1])==ARGV[1] then return redis.call('del',KEYS[1]) else return 0 end";
    await this.redis.eval(
      script,
      1,
      this.keys.taskLock(jobId, taskId),
      agentId,
    );
  }

  async completeTask(
    jobId: string,
    taskId: string,
    agentId: string,
    result = "",
  ): Promise<void> {
    await this.redis.hset(
      this.keys.taskHash(jobId, taskId),
      "status",
      "completed",
      "result",
      result,
      "owner",
      agentId,
    );
    await this.releaseTaskLock(jobId, taskId, agentId);

    // Auto-unblock downstream tasks.
    const downstream = await this.redis.smembers(
      this.keys.taskBlocks(jobId, taskId),
    );
    for (const tid of downstream) {
      await this.redis.srem(this.keys.taskBlockedBy(jobId, tid), taskId);
    }

    await this.redis.publish(
      "channel:tasks:completed",
      JSON.stringify({
        job_id: jobId,
        task_id: taskId,
        agent_id: agentId,
        status: "completed",
        result,
      }),
    );
    await this.publishJobEvent(jobId, {
      type: "task_completed",
      task_id: taskId,
      agent_id: agentId,
    });
  }

  async failTask(
    jobId: string,
    taskId: string,
    agentId: string,
    error: string,
  ): Promise<void> {
    await this.redis.hset(
      this.keys.taskHash(jobId, taskId),
      "status",
      "failed",
      "result",
      error,
      "owner",
      agentId,
    );
    await this.releaseTaskLock(jobId, taskId, agentId);

    await this.redis.publish(
      "channel:tasks:completed",
      JSON.stringify({
        job_id: jobId,
        task_id: taskId,
        agent_id: agentId,
        status: "failed",
        result: error,
      }),
    );
    await this.publishJobEvent(jobId, {
      type: "task_failed",
      task_id: taskId,
      agent_id: agentId,
      error,
    });
  }

  async getUnblockedTasks(jobId: string): Promise<TaskType[]> {
    const taskIds = await this.redis.zrange(
      this.keys.jobTasksZset(jobId),
      0,
      -1,
    );
    const unblocked: TaskType[] = [];
    for (const tid of taskIds) {
      const data = await this.redis.hgetall(
        this.keys.taskHash(jobId, tid),
      );
      if (!data || Object.keys(data).length === 0) continue;
      if (data.status !== "pending" || data.owner) continue;
      const blockers = await this.redis.smembers(
        this.keys.taskBlockedBy(jobId, tid),
      );
      if (blockers.length > 0) continue;
      const task = await this.getTask(jobId, tid);
      if (task) unblocked.push(task);
    }
    return unblocked;
  }

  // --- Agent registry (Hash + ZSet) ----------------------------------------

  async registerAgent(
    agentId: string,
    agentData: Record<string, string>,
  ): Promise<void> {
    const flat: string[] = [];
    for (const [k, v] of Object.entries(agentData)) {
      flat.push(k, v);
    }
    if (flat.length > 0) {
      await this.redis.hset(this.keys.agentHash(agentId), ...flat);
    }
    await this.redis.zadd(this.keys.agentsActive, Date.now(), agentId);
  }

  async heartbeat(
    agentId: string,
    status: string,
    currentTask = "",
  ): Promise<void> {
    const ts = String(Date.now() / 1000);
    await this.redis.zadd(this.keys.agentsActive, Date.now(), agentId);
    await this.redis.hset(
      this.keys.agentHash(agentId),
      "status",
      status,
      "current_task",
      currentTask,
      "heartbeat_ts",
      ts,
    );
  }

  async getIdleWorkers(opts?: {
    staleAfterSeconds?: number;
    jobId?: string;
  }): Promise<string[]> {
    const stale = opts?.staleAfterSeconds ?? 30;
    const cutoff = Date.now() - stale * 1000;
    const agentIds = await this.redis.zrangebyscore(
      this.keys.agentsActive,
      cutoff,
      "+inf",
    );
    const idle: string[] = [];
    for (const aid of agentIds) {
      const data = await this.redis.hgetall(this.keys.agentHash(aid));
      if (data.role !== "worker" || data.status !== "idle") continue;
      if (opts?.jobId && data.job_id !== opts.jobId) continue;
      idle.push(aid);
    }
    return idle;
  }

  async getStaleAgents(thresholdSeconds = 30): Promise<string[]> {
    const cutoff = Date.now() - thresholdSeconds * 1000;
    return this.redis.zrangebyscore(
      this.keys.agentsActive,
      "-inf",
      cutoff,
    );
  }

  async listAgents(): Promise<Record<string, string>[]> {
    const agentIds = await this.redis.zrange(
      this.keys.agentsActive,
      0,
      -1,
    );
    const agents: Record<string, string>[] = [];
    for (const aid of agentIds) {
      const data = await this.redis.hgetall(this.keys.agentHash(aid));
      agents.push(data);
    }
    return agents;
  }

  async listJobAgents(
    jobId: string,
    staleAfterSeconds = 30,
  ): Promise<string[]> {
    const cutoff = Date.now() - staleAfterSeconds * 1000;
    const agentIds = await this.redis.zrangebyscore(
      this.keys.agentsActive,
      cutoff,
      "+inf",
    );
    const result: string[] = [];
    for (const aid of agentIds) {
      const data = await this.redis.hgetall(this.keys.agentHash(aid));
      if (data.job_id !== jobId) continue;
      if (data.role !== "worker" && data.role !== "leader") continue;
      result.push(aid);
    }
    return result;
  }

  // --- Worker requests (List) ----------------------------------------------

  async submitWorkerRequest(request: WorkerRequestType): Promise<void> {
    await this.redis.lpush(
      this.keys.workerRequestQueue,
      JSON.stringify(request),
    );
  }

  async waitForWorkerRequest(
    timeout = 0,
  ): Promise<WorkerRequestType | null> {
    const result = await this.redis.brpop(
      this.keys.workerRequestQueue,
      timeout,
    );
    if (!result) return null;
    const [, raw] = result;
    return WorkerRequest.parse(JSON.parse(raw));
  }

  // --- Events --------------------------------------------------------------

  async publishJobEvent(
    jobId: string,
    event: Record<string, unknown>,
  ): Promise<void> {
    try {
      await this.redis.publish(
        this.keys.jobEventsChannel(jobId),
        JSON.stringify(event),
      );
    } catch {
      // Pub/Sub is best-effort.
    }
  }
}
